### PR TITLE
Extract incremental execution

### PR DESCRIPTION
**Pull request by lptr**
Move code related to incremental change detection to the `:execution` project.

### CODE REVIEW

**Comment by lptr** [comment]
@wolfs please take a look; it should already work. It's probably best to look commit-by-commit.

The execution engine now handles everything related to detecting incremental changes. It also includes actually storing the state itself after execution (the same way we do for tasks, i.e. we take into account successful execution and skip doing it if there are no output changes).

I'm planning to start dissecting the tests, so that each step can be tested on its own.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java [hunk 1]
```diff
@@ -16,17 +16,41 @@
...
+     * The outcome of executing the task.
+     */
+    Try<ExecutionOutcome> getOutcome();
+
     /**
      * If the execution resulted in some previous output being reused, this returns its origin metadata.
      */
     Optional<OriginMetadata> getReusedOutputOriginMetadata();
 
     TaskExecuterResult NO_REUSED_OUTPUT = new TaskExecuterResult() {
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java]: I think the name of this constant is a little bit strange. `NO_REUSED_OUTPUT` sounds like the task did not re-use any output. For me that means that the task executed. It also says above "An empty list means the task was not executed". This is in contradiction with the outcome "ExecutionOutcome.EXECUTED_FULLY".

I don't really understand what the executer result means :(

File: subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java [hunk 2]
```diff
@@ -17,7 +17,25 @@
...
+
+    /**
+     * The work has been executed with information about the changes that happened since the previous execution.
+     */
+    EXECUTED_INCREMENTALLY,
+
+    /**
+     * The work has been executed with no incremental change information.
+     */
+    EXECUTED_FULLY
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java]: How about we call this outcome `REBUILT`?

File: subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy [hunk 3]
```diff
@@ -167,15 +178,6 @@ class ExecutionTest extends Specification {
...
 
-        @Override
-        void persistResult(ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs, boolean successful, OriginMetadata originMetadata) {
-            throw new UnsupportedOperationException()
-        }
-
-        @Override
-        Optional<ExecutionStateChanges> getChangesSincePreviousExecution() {
-            throw new UnsupportedOperationException()
-        }
```
Comment (wolfs) [inline subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy]: Nit: Double newline.

File: subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy [hunk 4]
```diff
@@ -203,4 +205,32 @@ class ExecutionTest extends Specification {
         }
     }
 
+    def execute(Supplier<UnitOfWork> work) {
+        executionStep.execute(new IncrementalChangesContext() {
+            @Override
+            Optional<ExecutionStateChanges> getChanges() {
```
Comment (wolfs) [inline subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy]: I find it pretty annoying that those things need to be overridden together each time we create a context further down the road, only to pass things through. Do you think that maybe the `rebuildReason` and the execution states should be merged into one object?

File: subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/IncrementalExecutionTest.groovy [hunk 5]
```diff
@@ -57,13 +59,17 @@ import org.gradle.test.fixtures.file.TestFile
...
 import java.time.Duration
-import java.util.function.BooleanSupplier
+import java.util.function.Supplier
 
-import static org.gradle.internal.execution.ExecutionOutcome.EXECUTED
+import static org.gradle.internal.execution.ExecutionOutcome.EXECUTED_FULLY
 import static org.gradle.internal.execution.ExecutionOutcome.UP_TO_DATE
+import static org.gradle.internal.execution.history.changes.OutputFileChanges.OutputHandling.DETECT_ADDED
 
+// FIXME:lptr
```
Comment (wolfs) [inline subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/IncrementalExecutionTest.groovy]: Let's not forget to fix this test before we merge the PR.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java [hunk 6]
```diff
@@ -16,6 +16,6 @@
 
 package org.gradle.internal.execution;
 
-public interface WorkExecutor<R extends Result> {
-    R execute(UnitOfWork work);
+public interface WorkExecutor<C extends Context, R extends Result> {
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java]: Can we get rid of the type parameters of `WorkExecutor`? I don't see us supporting different types in the future, so why carry around the burden of the type parameters?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/WorkExecutorTestFixture.java [hunk 7]
```diff
@@ -70,7 +71,8 @@ public void close() {
     private BuildCancellationToken cancellationToken = new DefaultBuildCancellationToken();
     private final WorkExecutor<IncrementalContext, UpToDateResult> workExecutor;
 
-    WorkExecutorTestFixture(DefaultFileSystemMirror fileSystemMirror) {
+    WorkExecutorTestFixture(DefaultFileSystemMirror fileSystemMirror,
+                            ExecutionHistoryStore executionHistoryStore) {
```
Comment (wolfs) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/WorkExecutorTestFixture.java]: I guess the users of the test fixture need to be adjusted as well.

File: subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy (hunk 4 above)
Comment (lptr) [reply subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy]: The execution states each, and the rebuild reason again will be added in a separate `Step`, at least if we keep the original structure (which I intend to do at least first). I'd like to review these choices later, and improve on them once we have all the stuff we want moved over to the execution engine.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java (hunk 6 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java]: Yes, I think in the end we will be able to remove a bunch of type parameters. Again, I'd like to first move everything over, and then do a cleanup.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/WorkExecutorTestFixture.java (hunk 7 above)
Comment (lptr) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/WorkExecutorTestFixture.java]: Yep, done.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java (hunk 2 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java]: How about `EXECUTED`? `REBUILT` implies a previous execution, and I don't like to mix in the verb "build" here.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java]: OK, I simplified this by only providing information whether or not the task was executed incrementally, as this is the only thing we need to understand at the call site.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java [hunk 8]
```diff
@@ -262,34 +305,9 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
...
-                        outputFilesRepository.recordOutputs(finalOutputs.values());
-                    }
-                });
-            }
-        }
-
         private boolean hasAnyOutputFileChanges(ImmutableSortedMap<String, FileCollectionFingerprint> previous, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> current) {
             return !previous.keySet().equals(current.keySet())
-                || new OutputFileChanges(previous, current).hasAnyChanges();
+                || new OutputFileChanges(previous, current, IGNORE_ADDED).hasAnyChanges();
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: Well spotted. I've simplified this a little now I think it makes more sense.

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: I also renamed the enum values to `INCLUDE_ADDED` and `EXCLUDE_ADDED`, and improved their Javadocs.

File: subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy (hunk 3 above)
Comment (lptr) [reply subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy]: Removed.

**Comment by lptr** [comment]
@wolfs I addressed your comments, please take another look. Tomorrow I'll work on the tests. Perhaps we can merge this already if you're happy, so you can rebase the incremental transforms stuff on top of this.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: I saw you removed the enums again.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+public class ResolveChangesStep<R extends Result> implements Step<IncrementalContext, R> {
+    private final Step<? super IncrementalChangesContext, R> delegate;
+
+    public ResolveChangesStep(Step<? super IncrementalChangesContext, R> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public R execute(IncrementalContext context) {
+        final UnitOfWork work = context.getWork();
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java]: No need to use final here, right?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java [hunk 10]
```diff
@@ -63,10 +58,17 @@
      */
     Optional<? extends Iterable<String>> getChangingOutputs();
 
+    /**
+     * Whether files added are to be considered part of the output of the work.
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java]: I think it is not about whether files added to the output locations are considered an output of the work. When the task adds new files in an output directory, they are always considered an output of the work.

It is about whether the task is considered up-to-date if something else adds outputs to the output locations, without changing the recorded outputs of the last execution of the work. So it is about whether to allow overlapping outputs or not, right?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java (hunk 2 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java]: I find it strange that we call the outcome `EXECUTED` and have a `rebuildReason`, `RebuildExecutionStateChanges`, etc. Your point about "rebuild" implying a previous execution applies there as well. I think we should choose one name and stick with it.

So it seems that we use "incremental" to denote the incremental execution. We do that consistently. The question is what is the opposite of "incremental". Using executed here would mean that we use nothing (absence of incremental) to denote the non-incrementalness (that would imply `IncrementalExecutionStateChanges` vs. `DefaultExecutionStateChanges`). We could also use non-incremental to specify the non-incrementalness (e.g. `EXECUTED_NON_INCREMENTALLY`, `NonIncrementalExecutionStateChanges`, etc.).

We could try to use `full` execution as an opposite of incremental as well, though I think we then need to use it consistently.

Currently, I prefer to use non-incremental.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java]: I still don't understand what `NO_REUSED_OUTPUT` means, I think we should use a better name here. This executer result is not that the task didn't re-use any output. Any task that executes doesn't re-use the output. This is about the task not even trying to re-use the outputs.

How about `SKIPPED_WITHOUT_OUTPUTS` or just `WITHOUT_OUTPUTS`?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java (hunk 6 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java]: I don't quite understand why you introduced the `Context` type parameter in the first place. We can do the cleanup afterwards.

**Comment by wolfs** [comment]
@lptr I added some more comments and replied to yours. I think we can merge this PR (as long as tests are passing) and address the  remaining issues as a follow up PR.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java]: `WITHOUT_OUTPUTS` it is.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java (hunk 2 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java]: `EXECUTED_NON_INCREMENTALLY` it is.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java (hunk 9 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java]: Removed.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java (hunk 10 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java]: Clarified this a bit.

**Pull request merged**